docs/downloader.rst

```diff
@@ -0,0 +1,169 @@
+  Deletion Snapshots contain no such key.
+- ``metadata``: information about the Snapshot, a read-only Capability pointing to a JSON-serialized dict containing:
+  - ``snapshot_version``: int(1) currently
+  - ``name``: the name of this snapshot (a mangled relative path)
+  - ``author``: a dict containing:
+    - ``name``: arbitrary name
+    - ``verify_key``: base64-encoded public key of the author
+  - ``parents``: a list of immutable Capability-strings of any parent Snapshots
+  - (not yet) additionally, in the Tahoe metadata for this metadata-capability is a ``magic_folder`` dict with the following keys:
+    - ``author_signature``: base64-encoded signature which signs the content-Capability, metadata-Capability and name
+
+If there are zero parents this is the first version of a file.
+If there is one parent it is a modification.
+If there are two or more parents this version is the resolution of a conflict.
+
+When reading the code, there is a method ``create_snapshot_from_capability`` which downloads a capability-string from Tahoe and returns a ``RemoteSnapshot`` instance.
+
+
+General Operation
+-----------------
+
+The ``DownloaderService`` is responsible for deciding which Snapshots to download.
+The capability strings of Snapshots to download are given to ``RemoteSnapshotCacheService`` which downloads a Snapshot and all its relevant parents into memory.
+Once a Snapshot is downloaded it is passed to the ``MagicFolderUpdaterService`` which stages the actual content and determins if the change is an "overwrite" or a "conflict" and changes the local filesystem accordingly.
+
+Downloading new Snapshots from other participants ultimately causes changes to the local filesystem (in the magic-folder).
+
+
+What Snapshots to Download
+--------------------------
+
+Each configured magic-folder has a ``collective_dircap`` which is a Tahoe-LAFS Directory Capability ("dircap") for the list of participants.
+If this dircap is writable then this device is the administrator (and the only device which can modify the participants).
+
+In either case, the capability can be downloaded.
+It will be a Tahoe-LAFS directory containing a series of sub-directories; these are the participants.
+The directory name is the Participant name and points at a directory read-capability where all the files in their magic-folder are stored.
+This directory is known as the "Personal DMD" (where DMD stands for "Distributed Mutable Directory").
+
+The entries in a user's Personal DMD are flat (no subdirectories) and point a (mangled) relative path-name to a Snapshot.
+
+So, Snapshots to download are discovered by:
+
+- reading the Collective DMD
+- for each user in it:
+  - read their Personal DMD
+  - for each entry in the Personal DMD:
+    - queue the Snapshot for caching, including all parents
+    - after download, queue the Snapshot for updates
+
+Note that we don't always download *every* parent; we can stop early if a common ancestor is found.
```

Comment:
The code already does this.

Reply:
Then I think the discussion above should explain what we are actually doing.

Reply:
It is doing what is discussed here: stopping early if it already finds the common ancestor.

Reply:
- We should include that above, rather than saying we get all parents and then have a note afterwards saying we don't
- We should how we find a common ancestor.